### DISCUSSION TITLE
Make `event-log` can be turned off

### DISCUSSION THREAD

**Discussion by lorniu** [joaotavora/eglot#776] [Ideas]
I found `eglot` sometimes slower than `lsp-mode` when I edit Java or C# file.
The completions always delay about 1 seconds or more after every type, but in `lsp-mode` it's not.

I'm puzzled by this, and at last I found, it's the *event-log* which is turned on by default, that cause the issue.

There is no convenient way to turn off the event-log, so I have to advice the function, like this:
```elisp
(defvar jsonrpc-log-event-p nil)

(defun jsonrpc--log-event-advice (f &rest args)
  (if jsonrpc-log-event-p (apply f args)))

(advice-add #'jsonrpc--log-event :around #'jsonrpc--log-event-advice)
```

And then the performance improvement is huge, completions popup almost no delay after type.

I don't think it's nessasary to alway keep event-log on, but is there any way to turn it off conveniently?

Or is it necessary to add options to give users  to choose whether to turn on the event-log?

**Comment by joaotavora** [comment]
See if this helps:
```
eglot-events-buffer-size is a variable defined in `eglot.el'.

Its value is 2000000

Control the size of the Eglot events buffer.
If a number, don't let the buffer grow larger than that many
characters.  If 0, don't use an events buffer at all.  If nil,
let the buffer grow forever.

  You can customize this variable.

[back]
```

**Reply by lorniu** [reply]
Yes, it works. Thanks, I should read the code carefully.

**Reply by skangas** [reply]
Is it worth the performance hit to keep it on by default? The alternative would be to add instructions to the bug reporting page for how to turn it on before reporting a bug.

**Reply by Eason0210** [reply]
Hi @joaotavora ,

To improve performance, I tried to set `eglot-events-buffer-size `to` 0`, but the events buffer will still be created when I run <kbd>M-x</kbd> ` eglot `with `clangd`. Add got the content in *EGLOT events*  as below, and it will grow up when I input characters  in my` test.cpp` buffer:
``` bash
[stderr] I[09:11:38.996] Apple clangd version 13.1.6 (clang-1316.0.21.2.3)
[stderr] I[09:11:38.997] Features: mac+xpc
[stderr] I[09:11:38.997] PID: 17457
[stderr] I[09:11:38.997] Working directory: /Users/eason/Code/Cpp
[stderr] I[09:11:38.997] argv[0]: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clangd
[stderr] I[09:11:38.997] Starting LSP over stdin/stdout
[stderr] I[09:11:38.997] <-- initialize(1)
[stderr] I[09:11:39.002] --> reply:initialize(1) 4 ms
[stderr] I[09:11:39.003] <-- initialized
[stderr] I[09:11:39.004] <-- textDocument/didOpen
[stderr] I[09:11:39.005] Failed to find compilation database for /Users/eason/code/Cpp/test2.cpp
[stderr] I[09:11:39.005] ASTWorker building file /Users/eason/code/Cpp/test2.cpp version 0 with command clangd fallback
[stderr] [/Users/eason/code/Cpp]
[stderr] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang /Users/eason/code/Cpp/test2.cpp -resource-dir=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/13.1.6
[stderr] I[09:11:39.016] <-- workspace/didChangeConfiguration
[stderr] I[09:11:39.538] <-- textDocument/signatureHelp(2)
[stderr] I[09:11:39.538] <-- textDocument/hover(3)
[stderr] I[09:11:39.538] <-- textDocument/documentHighlight(4)
[stderr] I[09:11:39.570] --> reply:textDocument/signatureHelp(2) 31 ms
[stderr] I[09:11:39.581] --> textDocument/publishDiagnostics
[stderr] I[09:11:39.581] --> reply:textDocument/hover(3) 43 ms
[stderr] I[09:11:39.581] --> reply:textDocument/documentHighlight(4) 43 ms
```

But when I set `eglot-events-buffer-size`  to `1`, the *EGLOT events* buffer will be created with empty content, and will not grow up when in input characters  in my` test.cpp` buffer

I also tested with `python` with `pyright`, set to 0 will not create *EGLOT events* buffer. And I heard that other friends also tried on `css`、`java`，`Lua`, aslo no  *EGLOT events* buffer created with  `eglot-events-buffer-size`  set to `1` . 

So, looks like it only happened on` clangd `recently.

I am not sure if it is a bug or it's designed.

**Reply by joaotavora** [reply]
I doubt clang's stderr output that is creating any type of "performance problem", but if you want just reduce `clangd` s error reporting with something like `clangd --log=error`.  I use `clangd` as well and the slowdowns I notice, I think, are not related to logging at all, just to the inherent difficulty clangd has when parsing large C++ projects.

**Reply by Eason0210** [reply]
> I doubt clang's stderr output that is creating any type of "performance problem"

I also didn't experience performance issues with the stderr log of clangd. 
>  if you want just reduce clangd s error reporting with something like` clangd --log=error`

Thanks for mention this option. I will try it.

**Comment by lorniu** [comment]
One disadvantage of using `eglot-events-buffer-size` to do this is that, it can't turn on/off the event log on the fly, I have to restart eglot completely  to make the change take effect. Maybe using some dynamically-bind variable to control the output of jsonrpc-log is a better option.  

Advice on the method can make it work ok, but I don't think I am the only man need this, so, maybe a lighter way is needed.

```elisp
(defvar eglot-log-event-p nil)

(defun jsonrpc--log-event$toggle-event-log (f &rest args)
  (when (and eglot-log-event-p
             (ignore-errors
               (eq (type-of (car args)) 'eglot-lsp-server)))
    (apply f args)))

(advice-add #'jsonrpc--log-event :around #'jsonrpc--log-event$toggle-event-log)

;; Toggle:

(defun im/toggle-eglot-event-log ()
  (interactive)
  (setq eglot-log-event-p (not eglot-log-event-p))
  (message "EGLOG event log is current: %s"
           (if eglot-log-event-p "ON" "OFF")))
```


Thanks for your amazing work.

**Reply by joaotavora** [reply]
I just noticed this here after a long time, sorry about that.  It's true you need to restart the connection for the variable to take effect, but it's also true that that is as easy as `M-x eglot-reconnect`.

**Reply by casouri** [reply]
I would explain that in the doctoring. I just set the variable and was surprised that it has no effect.

**Comment by joaotavora** [comment]
Yuan Fu ***@***.***> writes:

> I would explain that in the doctoring. I just set the variable and was
> surprised that it has no effect.

Sounds reasonable.  I'll make the change unless you beat me to it with a
PR.  You've got a copyright assignment, don't you?

João

**Reply by casouri** [reply]
I do ;-) I can do it (in a couple of days).
